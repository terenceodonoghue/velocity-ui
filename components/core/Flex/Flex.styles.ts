import { css } from '@emotion/react';

const flex = css`
  display: flex;
`;

export default flex;
