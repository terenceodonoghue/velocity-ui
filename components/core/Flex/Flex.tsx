import React, { FunctionComponent } from 'react';
import flex from './Flex.styles';
import type { FlexProps } from './Flex.types';

const Flex: FunctionComponent<FlexProps> = ({
  alignItems,
  basis,
  children,
  direction,
  grow,
  justifyContent,
  shrink,
  wrap,
  ...props
}) => (
  <div
    css={[
      flex,
      {
        alignItems,
        justifyContent,
        flex:
          (grow >= 0 || shrink >= 0 || basis || basis >= 0) &&
          `${grow ?? 0} ${shrink ?? 1} ${basis ?? 'auto'}`,
        flexDirection: direction,
        flexWrap: wrap === true ? 'wrap' : wrap,
      },
    ]}
    data-testid="flex"
    {...props}
  >
    {children}
  </div>
);

export default Flex;
