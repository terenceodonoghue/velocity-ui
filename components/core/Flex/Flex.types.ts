import { HTMLAttributes, ReactText } from 'react';

type AlignItemsProperty =
  // Basic keywords
  | 'normal'
  | 'stretch'
  // Position alignment
  | 'center'
  | 'start'
  | 'end'
  | 'flex-start'
  | 'flex-end'
  // Baseline alignment
  | 'baseline'
  | 'first baseline'
  | 'last baseline'
  | 'safe center'
  | 'unsafe center'
  // Global values
  | 'inherit'
  | 'initial'
  | 'unset';

type JustifyContentProperty =
  // Basic keywords
  | 'normal'
  | 'stretch'
  // Position alignment
  | 'center'
  | 'start'
  | 'end'
  | 'flex-start'
  | 'flex-end'
  | 'left'
  | 'right'
  // Distributed alignment
  | 'space-between'
  | 'space-around'
  | 'space-evenly'
  // Overflow alignment
  | 'safe center'
  | 'unsafe center'
  // Global values
  | 'inherit'
  | 'initial'
  | 'unset';

type FlexBasisProperty = ReactText;

type FlexDirectionProperty = 'column' | 'row';

type FlexWrapProperty = 'nowrap' | 'wrap' | 'wrap-reverse' | true;

export interface FlexProps extends HTMLAttributes<HTMLDivElement> {
  alignItems?: AlignItemsProperty;
  basis?: FlexBasisProperty;
  direction?: FlexDirectionProperty;
  grow?: number;
  justifyContent?: JustifyContentProperty;
  shrink?: number;
  wrap?: FlexWrapProperty;
}
