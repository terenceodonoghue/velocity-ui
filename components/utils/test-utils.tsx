import React, { FunctionComponent, ReactElement } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import ThemeProvider from '~/components/core/ThemeProvider';

const Providers: FunctionComponent = ({ children }) => {
  return <ThemeProvider>{children}</ThemeProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

export { customRender as render };
