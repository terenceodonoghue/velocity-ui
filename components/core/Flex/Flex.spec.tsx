import React from 'react';
import faker from 'faker';
import { render } from 'test-utils';
import Flex from './Flex';

describe('Flex', () => {
  it('renders content', () => {
    const content = faker.hacker.phrase();
    const { container, getByTestId } = render(<Flex>{content}</Flex>);

    expect(getByTestId('flex')).toHaveTextContent(content);
    expect(container).toMatchSnapshot();
  });

  it('has align-items propery', () => {
    const { container, getByTestId } = render(<Flex alignItems="flex-start" />);

    expect(getByTestId('flex')).toHaveStyle({ alignItems: 'flex-start' });
    expect(container).toMatchSnapshot();
  });

  it('has align-items propery', () => {
    const { container, getByTestId } = render(
      <Flex justifyContent="flex-start" />,
    );

    expect(getByTestId('flex')).toHaveStyle({ justifyContent: 'flex-start' });
    expect(container).toMatchSnapshot();
  });

  it('has flex-grow property', () => {
    const { container, getByTestId } = render(<Flex grow={1} />);

    expect(getByTestId('flex')).toHaveStyle({ flex: '1 1 auto' });
    expect(container).toMatchSnapshot();
  });

  it('has flex-shrink property', () => {
    const { container, getByTestId } = render(<Flex shrink={0} />);

    expect(getByTestId('flex')).toHaveStyle({ flex: '0 0 auto' });
    expect(container).toMatchSnapshot();
  });

  it('has flex-basis property', () => {
    const { container, getByTestId } = render(<Flex basis="50%" />);

    expect(getByTestId('flex')).toHaveStyle({ flex: '0 1 50%' });
    expect(container).toMatchSnapshot();
  });

  it('has flex-direction property', () => {
    const { container, getByTestId } = render(<Flex direction="column" />);

    expect(getByTestId('flex')).toHaveStyle({ flexDirection: 'column' });
    expect(container).toMatchSnapshot();
  });

  it('has flex-wrap property', () => {
    const { container, getByTestId, rerender } = render(<Flex wrap="nowrap" />);

    expect(getByTestId('flex')).toHaveStyle({ flexWrap: 'nowrap' });
    expect(container).toMatchSnapshot();

    rerender(<Flex wrap />);

    expect(getByTestId('flex')).toHaveStyle({ flexWrap: 'wrap' });
    expect(container).toMatchSnapshot();
  });
});
