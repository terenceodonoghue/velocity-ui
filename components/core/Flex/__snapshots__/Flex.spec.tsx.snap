// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex has align-items propery 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

<div>
  <div
    class="emotion-0"
    data-testid="flex"
  />
</div>
`;

exports[`Flex has align-items propery 2`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
}

<div>
  <div
    class="emotion-0"
    data-testid="flex"
  />
</div>
`;

exports[`Flex has flex-basis property 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 50%;
  -ms-flex: 0 1 50%;
  flex: 0 1 50%;
}

<div>
  <div
    class="emotion-0"
    data-testid="flex"
  />
</div>
`;

exports[`Flex has flex-direction property 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="emotion-0"
    data-testid="flex"
  />
</div>
`;

exports[`Flex has flex-grow property 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

<div>
  <div
    class="emotion-0"
    data-testid="flex"
  />
</div>
`;

exports[`Flex has flex-shrink property 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

<div>
  <div
    class="emotion-0"
    data-testid="flex"
  />
</div>
`;

exports[`Flex has flex-wrap property 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

<div>
  <div
    class="emotion-0"
    data-testid="flex"
  />
</div>
`;

exports[`Flex has flex-wrap property 2`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

<div>
  <div
    class="emotion-0"
    data-testid="flex"
  />
</div>
`;

exports[`Flex renders content 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <div
    class="emotion-0"
    data-testid="flex"
  >
    The SMS program is down, synthesize the redundant feed so we can synthesize the SQL hard drive!
  </div>
</div>
`;
